File: core/src/main/java/com/oneandone/typedrest/ElementEndpointImpl.java
package com.oneandone.typedrest;

import java.io.*;
import java.net.*;
import java.util.Optional;
import lombok.Getter;
import org.apache.http.*;
import org.apache.http.client.fluent.*;
import org.apache.http.entity.*;
import org.apache.http.util.*;

/**
 * REST endpoint that represents a single <code>TEntity</code>.
 *
 * @param <TEntity> The type of entity the endpoint represents.
 */
public class ElementEndpointImpl<TEntity>
        extends AbstractETagEndpoint implements ElementEndpoint<TEntity> {

    @Getter
    private final Class<TEntity> entityType;

    /**
     * Creates a new element endpoint.
     *
     * @param referrer The endpoint used to navigate to this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>referrer</code>'s.
     * @param entityType The type of entity the endpoint represents.
     */
    public ElementEndpointImpl(Endpoint referrer, URI relativeUri, Class<TEntity> entityType) {
        super(referrer, relativeUri);
        this.entityType = entityType;
    }

    /**
     * Creates a new element endpoint.
     *
     * @param referrer The endpoint used to navigate to this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>referrer</code>'s. Prefix <code>./</code> to append a trailing
     * slash to the <code>referrer</code> URI if missing.
     * @param entityType The type of entity the endpoint represents.
     */
    public ElementEndpointImpl(Endpoint referrer, String relativeUri, Class<TEntity> entityType) {
        super(referrer, relativeUri);
        this.entityType = entityType;
    }

    @Override
    public TEntity read()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException {
        HttpEntity content = getContent();
        return serializer.readValue(EntityUtils.toString(content), entityType);
    }

    @Override
    public boolean exists()
            throws IOException, IllegalAccessException {
        try {
            executeAndHandle(Request.Head(uri));
        } catch (FileNotFoundException ex) {
            return false;
        }
        return true;
    }

    @Override
    public Optional<Boolean> isSetAllowed() {
        return isVerbAllowed("PUT");
    }

    @Override
    public TEntity set(TEntity entity)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException {
        if (entity == null) {
            throw new IllegalArgumentException("entity must not be null.");
        }

        HttpEntity content = new StringEntity(serializer.writeValueAsString(entity), ContentType.APPLICATION_JSON);
        HttpResponse response = putContent(content);
        return (response.getEntity() == null)
                ? null
                : serializer.readValue(EntityUtils.toString(response.getEntity()), entityType);
    }

    @Override
    public Optional<Boolean> isDeleteAllowed() {
        return isVerbAllowed("DELETE");
    }

    @Override
    public void delete()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException {
        deleteContent();
    }
}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/generic/ElementEndpointImpl.kt
package net.typedrest.endpoints.generic

import net.typedrest.endpoints.*
import net.typedrest.errors.*
import net.typedrest.http.*
import okhttp3.*
import java.net.URI

/**
 * Endpoint for an individual resource.
 *
 * @param referrer The endpoint used to navigate to this one.
 * @param relativeUri The URI of this endpoint relative to the [referrer]'s.
 * @param entityType The type of entity the endpoint represents.
 * @param TEntity The type of entity the endpoint represents.
 */
open class ElementEndpointImpl<TEntity>(
    referrer: Endpoint,
    relativeUri: URI,
    private val entityType: Class<TEntity>
) : AbstractCachingEndpoint(referrer, relativeUri), ElementEndpoint<TEntity> {
    /**
     * Creates a new element endpoint.
     *
     * @param referrer The endpoint used to navigate to this one.
     * @param relativeUri The URI of this endpoint relative to the [referrer]'s. Add a `./` prefix here to imply a trailing slash on referrer's URI.
     * @param entityType The type of entity the endpoint represents.
     */
    constructor(referrer: Endpoint, relativeUri: String, entityType: Class<TEntity>) :
        this(referrer, URI(relativeUri), entityType)

    override val response: TEntity?
        get() = responseCache?.getBody()?.let { deserialize(it, entityType) }

    override fun read(): TEntity =
        getContent()?.let { deserialize(it, entityType) }
            ?: throw NotFoundException("Result not deserializable as ${entityType.simpleName}")

    override fun exists(): Boolean =
        httpClient.newCall(Request.Builder().head().uri(uri).build()).execute().use { response ->
            when {
                response.isSuccessful -> true
                response.code == HttpStatusCode.NotFound.code || response.code == HttpStatusCode.Gone.code -> false
                else -> {
                    errorHandler.handle(response)
                    false
                }
            }
        }

    override val isSetAllowed: Boolean?
        get() = isMethodAllowed(HttpMethod.PUT)

    override fun set(entity: TEntity): TEntity? =
        tryReadAs(putContent(serialize(entity, entityType)))

    override val isMergeAllowed: Boolean?
        get() = isMethodAllowed(HttpMethod.PATCH)

    override fun merge(entity: TEntity): TEntity? {
        responseCache = null
        return tryReadAs(execute(Request.Builder().patch(serialize(entity, entityType)).uri(uri).build()))
    }

    override val isDeleteAllowed: Boolean?
        get() = isMethodAllowed(HttpMethod.DELETE)

    override fun delete() = deleteContent()

    override fun update(updateAction: (TEntity) -> TEntity, maxRetries: Int): TEntity? {
        var retryCounter = 0
        while (true) {
            val entity = updateAction(read())

            try {
                return set(entity)
            } catch (ex: HttpException) {
                if (retryCounter++ >= maxRetries) throw ex
                ex.retryAfter?.let(Thread::sleep)
            }
        }
    }

    private fun tryReadAs(response: Response): TEntity? {
        if (response.code == HttpStatusCode.NoContent.code) {
            return null
        }

        return try {
            deserialize(response.body, entityType)
        } catch (ex: Exception) {
            null
        }
    }
}
